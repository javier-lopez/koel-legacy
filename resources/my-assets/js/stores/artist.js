/* eslint camelcase: ["error", {properties: "never"}] */

import Vue from 'vue'
import { union, difference, take, orderBy } from 'lodash'

import stub from '@/stubs/artist'

const UNKNOWN_ARTIST_ID = 1
const VARIOUS_ARTISTS_ID = 2

export const artistStore = {
  stub,
  cache: [],

  state: {
    artists: []
  },

  init (artists) {
    this.all = artists

    // Traverse through artists array to get the cover and number of songs for each.
    this.all.forEach(artist => this.setupArtist(artist))
  },

  setupArtist (artist) {
    Vue.set(artist, 'playCount', 0)
    Vue.set(artist, 'info', null)
    Vue.set(artist, 'albums', [])
    Vue.set(artist, 'songs', [])

    this.cache[artist.id] = artist

    return artist
  },

  get all () {
    return this.state.artists
  },

  set all (value) {
    this.state.artists = value
  },

  byId (id) {
    return this.cache[id]
  },

  /**
   * Adds an artist/artists into the current collection.
   *
   * @param  {Array.<Object>|Object} artists
   */
  add (artists) {
    [].concat(artists).forEach(artist => {
      this.setupArtist(artist)
      artist.playCount = artist.songs.reduce((count, song) => count + song.playCount, 0)
    })

    this.all = union(this.all, artists)
  },

  purify () {
    this.compact()
  },

  /**
   * Remove empty artists from the store.
   */
  compact () {
    const emptyArtists = this.all.filter(artist => artist.songs.length === 0)
    if (!emptyArtists.length) {
      return
    }

    this.all = difference(this.all, emptyArtists)
    emptyArtists.forEach(artist => delete this.cache[artist.id])
  },

  isVariousArtists: artist => artist.id === VARIOUS_ARTISTS_ID,

  isUnknownArtist: artist => artist.id === UNKNOWN_ARTIST_ID,

  getSongsByArtist: artist => artist.songs,

  getMostPlayed (n = 6) {
    // Only non-unknown artists with actually play count are applicable.
    // Also, "Various Artists" doesn't count.
    const applicable = this.all.filter(artist => {
      return artist.playCount &&
        !this.isUnknownArtist(artist) &&
        !this.isVariousArtists(artist)
    })

    return take(orderBy(applicable, 'playCount', 'desc'), n)
  }
}
