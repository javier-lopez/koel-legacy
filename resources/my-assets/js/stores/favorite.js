import { difference, union } from 'lodash'
import NProgress from 'nprogress'

import { http } from '@/services'
import { alerts, pluralize } from '@/utils'

export const favoriteStore = {
  state: {
    songs: [],
    length: 0,
    fmtLength: ''
  },

  get all () {
    return this.state.songs
  },

  set all (value) {
    this.state.songs = value
  },

  toggleOne (song) {
    // Don't wait for the HTTP response to update the status, just toggle right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    song.liked = !song.liked
    song.liked ? this.add(song) : this.remove(song)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/like', { song: song.id }, ({ data }) => {
        // We don't really need to notify just for one song.
        resolve(data)
      }, error => reject(error))
    })
  },

  /**
   * Add a song/songs into the store.
   *
   * @param {Array.<Object>|Object} songs
   */
  add (songs) {
    this.all = union(this.all, [].concat(songs))
  },

  /**
   * Remove a song/songs from the store.
   *
   * @param {Array.<Object>|Object} songs
   */
  remove (songs) {
    this.all = difference(this.all, [].concat(songs))
  },

  clear () {
    this.all = []
  },

  like (songs) {
    // Don't wait for the HTTP response to update the status, just set them to Liked right away.
    // This may cause a minor problem if the request fails somehow, but do we care?
    songs.forEach(song => { song.liked = true })
    this.add(songs)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/batch/like', { songs: songs.map(song => song.id) }, ({ data }) => {
        alerts.success(`Added ${pluralize(songs.length, 'song')} into Favorites.`)
        resolve(data)
      }, error => reject(error))
    })
  },

  unlike (songs) {
    songs.forEach(song => { song.liked = false })
    this.remove(songs)

    NProgress.start()

    return new Promise((resolve, reject) => {
      http.post('interaction/batch/unlike', { songs: songs.map(song => song.id) }, ({ data }) => {
        alerts.success(`Removed ${pluralize(songs.length, 'song')} from Favorites.`)
        resolve(data)
      }, error => reject(error))
    })
  }
}
