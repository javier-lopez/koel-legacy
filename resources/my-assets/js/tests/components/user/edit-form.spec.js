import Component from '@/components/user/edit-form.vue'
import { userStore } from '@/stores'
import factory from '@/tests/factory'
import { mock } from '@/tests/__helpers__'

describe('components/user/edit-form', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('opens', () => {
    const user = factory('user', {
      name: 'Bob Dylan',
      email: 'knocking@heaven.door'
    })
    const wrapper = shallow(Component, {
      propsData: {
        user
      }
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('saves', () => {
    const user = factory('user')
    const updateStub = mock(userStore, 'update')
    const wrapper = shallow(Component, {
      propsData: {
        user
      }
    })
    wrapper.submit('form')
    expect(updateStub).toHaveBeenCalledWith(user, user.name, user.email, user.password)
  })

  it('cancels', () => {
    const user = factory('user')
    const updateStub = mock(userStore, 'update')
    const wrapper = shallow(Component, {
      propsData: {
        user
      }
    })
    wrapper.click('.btn-cancel')
    expect(wrapper.hasEmitted('close')).toBe(true)
    expect(updateStub).not.toHaveBeenCalled()
  })
})
