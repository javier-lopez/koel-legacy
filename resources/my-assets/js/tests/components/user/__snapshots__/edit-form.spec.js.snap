// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/user/edit-form opens 1`] = `
<div class="edit-user">
  <form class="user-edit">
    <header>
      <h1>Edit User</h1>
    </header>
    <div>
      <div class="form-row"><label>Name</label> <input title="Name" type="text" name="name" required="required"></div>
      <div class="form-row"><label>Email</label> <input title="Email" type="email" name="email" required="required"></div>
      <div class="form-row"><label>Password</label> <input type="password" name="password" placeholder="Leave blank for no changes"></div>
    </div>
    <footer><button type="submit" class="btn btn-green btn-update">Update</button> <button class="btn btn-white btn-cancel">Cancel</button></footer>
  </form>
</div>
`;
